app/frontend/src/components/ContributorForm/ContributorForm.tsx

```diff
@@ -103,7 +103,13 @@ export default function ContributorForm({
         .setExperience(data.experience)
         .setContribute(contribute)
         .setContributeWaysList(
-          data.contributeWays.split(",").filter((v) => !!v)
+          Object.entries(data.contributeWays).reduce<string[]>(
+            //contributeWays is an object of "ways" as keys, and "checked" booleans as values
+            //this reduces it to an array of the "ways" which were keys with "true" as a value
+            (previous, current) =>
+              current[1] ? [...previous, current[0]] : previous,
+            []
```

Comment:
How is this that much faster than what we had before? It's still going through the list and building up a list each time a checkbox is marked/unmarked so the code still has a linear runtime?

I reckon things might run a bit quicker if we just let checkboxes be its own individual fields and collect the results at the end at `onSubmit`, at least then this linear time (apparently slow) operation only runs once. We are basically abusing this one field to do multiple things when web forms aren't really designed for that and we shouldn't really be coupling the UI against the backend data structure. 

Otherwise, if we are commenting out that UI interaction in the test, I think the old code is a bit easier to read so maybe best to keep that?

Reply:
I don't think performance is the issue for this one, I don't see how this nor the previous version are particularly taxing. I thought it might be some obscure thing that would be eliminated by refactoring, and indeed it did seem to fix the test in the commit this first appeared, but didn't anymore when I applied it to develop. 

Anyway I prefer doing a map than explicitly listing every checkbox, but if you think the previous one is cleared we could put it back. (I personally prefer it this way, react-hook-form intends form fields to be used like this imo)

Reply:
>  I don't see how this nor the previous version are particularly taxing.

I thought the same and so didn't mention it the first time I reviewed it as I didn't want to pre-optimise things too much... until the tests now somehow choke on this so it is too slow? 😅 

> react-hook-form intends form fields to be used like this imo

I'm not sure I agree with that, especially after looking at their more [complex example with double level nested fields](https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-m8w6j?file=/src/index.js), they still register values into individual fields in a nested data structure rather than collapsing everything into one.

The performance of react-hook-form comes from preventing unnecessary re-renders from state updates from React normally, but it still runs `onChange` on every change so ideally the operation is a simple one, which is usually the case in most use cases... until we broke it by wrangling multiple fields into one with "complex" logic.

Reply:
> until we broke it by wrangling multiple fields into one
This version is multiple fields though - there's a controller for each checkbox so I suppose only one will re-render at a time. Before it was all in one controller.

I don't love this reduce but I don't think it's more opaque than the logic that was previously in `toggleCheckbox`. Some similar logic will still be necessary even if there was no `map`. (Converting true/false fields into a list of strings.)

Do you want me to explicitly list every checkbox? (And change `CONTRIBUTE_WAYS_OPTIONS` to individual strings?)

Reply:
My bad, I misread it completely and thought it was still like the old code where everything was in one field 😅 

In that case, no, there won't be much point explicitly listing the checkboxes as the resultant code will be the same 👍

Reply:
```suggestion
          Object.entries(data.contributeWays).reduce<string[]>(
            //contributeWays is an object of "ways" as keys, and "checked" booleans as values
            //this reduces it to an array of the "ways" which were keys with "true" as a value
            (previous, [contributeWay, checked]) =>
              checked ? [...previous, contributeWay] : previous,
            []
```
Minor nitpick, but I like having explicit variable names to make things clearer over index access